cmd/snap/cmd_run.go

```diff
@@ -1077,6 +1078,66 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
 		return env.ForExecEscapeUnsafe(snapenv.PreservedUnsafePrefix)
 	}
 
+	// Systemd automatically places services under a unique cgroup encoding the
+	// security tag but for apps and hooks we need to create a transient scope
+	// with similar purpose.
+	//
+	// The way this happens is as follows:
+	//
+	// 1) Services are implemented using systemd service units. Starting an
+	// unit automatically places it in a cgroup named after the service unit
+	// name. Snapd controls the name of the service units thus indirectly
+	// controls the cgroup name.
+	//
+	// 2) Non-services, including hooks, are started inside systemd
+	// transient scopes. Scopes are a systemd unit type that are defined
+	// programmatically and are meant for groups of processes started and
+	// stopped by an _arbitrary process_ (ie, not systemd). Systemd
+	// requires that each scope is given a unique name. We employ a scheme
+	// where random UUID is combined with the name of the security tag
+	// derived from snap application or hook name. Multiple concurrent
+	// invocations of "snap run" will use distinct UUIDs.
+	//
+	// Transient scopes allow launched snaps to integrate into
+	// the systemd design. See:
+	// https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/
+	//
+	// Programs running as root, like system-wide services and programs invoked
+	// using tools like sudo are placed under system.slice. Programs running as
+	// a non-root user are placed under user.slice, specifically in a scope
+	// specific to a logind session.
+	//
+	// This arrangement allows for proper accounting and control of resources
+	// used by snap application processes of each type.
+	//
+	// For more information about systemd cgroups, including unit types, see:
+	// https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/
+	needsTracking := true
+	if app := info.Apps[snapApp]; hook == "" && app != nil && app.IsService() {
+		// If we are not running a service app then we do not need to use
+		// application tracking. Services, both in the system and user scope,
+		// do not need tracking because systemd already places them in a
+		// tracking cgroup, named after the systemd unit name, and those are
+		// sufficient to identify both the snap name and the app name.
+		needsTracking = false
+	}
+	// Track, or confirm existing tracking from systemd.
+	var trackingErr error
+	if needsTracking {
+		trackingErr = cgroupCreateTransientScopeForTracking(securityTag)
```

Comment:
Note that internally we check the feature flag, so that this is a no-op until the flag is set.